R/general.R

```diff
@@ -48,3 +48,582 @@ remove_special_characters  <- function(x = "spe;c.l-cha,ac?ers"){
+#'               Enter UT030 if  county is Unknown.
+#'               Enter UT097 if student is Out of State, Out of US.
+#'               Enter UT099 if student is Out  of State, In the US.
+#'               This element should be logically consistent with S-11: S_STATE_ORIGIN and S 27: S_COUNTRY_ORIGIN.
+#'
+#' @importFrom magrittr %>%
+#' @importFrom dplyr mutate
+#' @importFrom dplyr select
+#'
+#' @param input_df A Data Frame. Must contain the following data fields: (first_admit_county_code, first_admit_state_code, first_admit_country_iso_code).
+#'
+#'
+#' @return Original data frame, with USHE data element s_10 appended.
+#' @export
+#'
+#' @examples
+#' gen_ushe_county_origin()
+#'
+gen_ushe_county_origin <- function(input_df=usheUtils::fake_student_df) {
+
+  output_df <- input_df %>%
+    # Calculate intermediate fields
+    mutate(county_origin = case_when(
+      ( first_admit_state_code == "UT" & !is.na(first_admit_county_code) ) ~ paste0("UT", first_admit_county_code),
+      first_admit_country_iso_code == "US" ~ "UT099",
+      first_admit_country_iso_code != "US" ~ "UT097",
+      TRUE ~ "UT030")) %>%
+    # Append USHE data element gen_ushe_county_origin
+    mutate( gen_ushe_county_origin = county_origin,
+            s_10 = county_origin,
+            g_04 = county_origin)
+
+  return(output_df)
+}
+
+#' @rdname gen_ushe_county_origin
+#' @examples s_10()
+#' @export
+s_10 <- gen_ushe_county_origin
+
+#' @rdname gen_ushe_county_origin
+#' @examples g_04()
+#' @export
+g_04 <- gen_ushe_county_origin
+
+#' Calculate USHE Element Social Security Number or Institutionally Assigned ID (Student ID)
+#'
+#' @details
+#'
+#' **USHE Documentation**
+#' - ELEMENT NAME: Institutionally Assigned ID
+#' - FIELD NAME: s_id, sc_id, and g_id
+#' - FIELD FORMAT: Varchar, 9 Characters
+#' - DEFINITION: The unique institutionally assigned identification number for each student intended to be used in lieu of using a student’s social security number.
+#'               G_Banner_ID must begin with an Alpha character representing the institution.
+#'
+#' @importFrom magrittr %>%
+#' @importFrom dplyr mutate
+#' @importFrom dplyr select
+#' @importFrom dplyr if_else
+#' @importFrom stringr str_replace_all
+#' @importFrom stringr str_starts
+#' @importFrom stringr str_length
+#'
+#' @param input_df A Data Frame. Must contain the following data fields: (student_id, ssn).
+#'
+#'
+#' @return Original data frame, with USHE data element s_03 appended.
+#' @export
+#'
+#' @examples
+#' gen_ushe_id()
+#'
+gen_ushe_id <- function(input_df=usheUtils::fake_student_df) {
+
+  output_df <- input_df %>%
+    # Calculate intermediate fields
+    # Note: ssn that start with 9 are actually IRS issued tax IDs; USHE only wants valid ssn.
+    mutate( id_intermediate = str_replace_all(ssn, "-", "")) %>%
+    mutate( id = if_else(is.na(id_intermediate) |
```

Comment:
We should write a function for this and use it in s_04 to make sure they match up. I can see this biting us later, if we don't address it now. I imagine the function being something like is_valid_ssn. And we would just use that in the call to if_else (in this function as well as s_04).